Print Roster

silverailscolo opened the pull request:
i18n and Dutch, German translations for Ops and Roster
Adjusted 1 test line

silverailscolo commented:
Travis Linuh headless failed
```
433 Still running (40 of 40): ./scripts/travis.sh
434 Timeout (40 minutes) reached. Terminating "./scripts/travis.sh"
```

rhwood commented on default.lcf:
```diff
@@ -108,3 +108,4 @@ log4j.category.javax.jmdns=OFF
 
 #log4j.category.jmri.jmrit.display.layoutEditor=DEBUG
 #log4j.category.jmri.jmrit.display.layoutEditor.LayoutBlock=DEBUG
+log4j.category.jmri.jmrit.roster.PrintRosterEntry=DEBUG
```
Can you please revert this change? I'm assuming this was committed in error (I make mistakes like this frequently myself as well).

silverailscolo replied:
Indeed, I left that in by mistake.

rhwood commented on java/src/jmri/jmrit/roster/PrintRosterEntry.java:
```diff
@@ -59,9 +68,10 @@ public PrintRosterEntry(RosterEntry rosterEntry, JmriJFrame parent, String filen
             }
             if ((base = root.getChild("programmer")) == null) {
-                log.error("xml file top element is not programmer");
+                log.error("xml file top element is not a programmer");
```
In this case we are looking for an element named "programmer" (wrapping it in quotes would be more clarifying than adding the indefinite article).

rhwood commented on java/src/jmri/jmrit/roster/PrintRosterEntry.java:
```diff
@@ -167,6 +179,15 @@ public boolean isBusy() {
+     * @param paneList list of programmer tabs
+     * @param flPane pane w/checkbox to select printing of "Function List"
+     * @param rMPane pane containing roster media (image)
+     * @param parent Roster JFrame calling this method
```
The parent parameter is not necessarily a Roster JFrame, or the caller of this method; its the window over which this dialog will be centered.

BTW, thank you for adding Javadoc comments. Its very helpful.

rhwood commented on java/src/jmri/jmrit/symbolicprog/PrintCvAction.java:
```diff
@@ -111,9 +111,13 @@ public void actionPerformed(ActionEvent e) {
-            // print a simple heading
-            s = "                 Value                   Value                   Value\n";
+            // print a simple heading with I18N
+            s = String.format("%1$21s", Bundle.getMessage("Value")) + String.format("%1$24s", Bundle.getMessage("Value")) +
```
You probably want to do this as a single format String:
```
s = String.format("%1$21s%1$24s%1$24s", Bundle.getMessage("Value"));
```
if that doesn't work, you can use:
```
s = String.format("%1$21s%2$24s%3$24s", Bundle.getMessage("Value"), Bundle.getMessage("Value"), Bundle.getMessage("Value"));
```

silverailscolo replied:
Approach #1 worked fine, thanks.

rhwood commented:
There are also some Javadoc issues flagged by the CI servers:
```
  [javadoc] Generating Javadoc
  [javadoc] Javadoc execution
  [javadoc] /home/travis/build/JMRI/JMRI/java/src/jmri/jmrit/roster/PrintRosterEntry.java:48: warning: no description for @param
  [javadoc]      * @param parent
  [javadoc]        ^
  [javadoc] /home/travis/build/JMRI/JMRI/java/src/jmri/jmrit/roster/PrintRosterEntry.java:49: warning: no description for @param
  [javadoc]      * @param filename
  [javadoc]        ^
  [javadoc] /home/travis/build/JMRI/JMRI/java/src/jmri/jmrit/roster/RosterEntry.java:190: error: @param name not found
  [javadoc]      * @param fileName xml file name for the user's Roster entry
  [javadoc]               ^
  [javadoc] 1 error
  [javadoc] 2 warnings
```

rhwood commented on java/src/jmri/jmrit/roster/PrintRosterEntry.java:
```diff
@@ -45,8 +45,8 @@
-     * @param parent
-     * @param filename
+     * @param parent window over which this dialog will be centered
+     * @param xml file name for the user's Roster entry
```
This is missing the name of the parameter.

silverailscolo commented:
Wrote in Javadoc how I understand this part to work. Also added Print (preview) to the Roster list context menu (the easy part). Bonus: I18N of the print out, except for the (Basic) print from the main DP roster view as that uses an older symbolicprog HardcopyWriter that won't allow for localizing the name of a pane/tab and still retrieve the contents from a decoder programmer pane. Requested changes by @rhwood adopted. Cleaned up DP3 nmra xml labels; please keep default (English) at the top of each block.
